File: SQL-AI/ProgrammingLanguage/Leetcode-Implementation/CPP/Permutations.cpp
class Solution {
public:
    vector<vector<int> > permute(vector<int> &num) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        vector<vector<int>> permutations;
        PermuteHelp( num, 0, permutations);
        return permutations;
    }
    
    void PermuteHelp( vector<int> &num, int offset, vector<vector<int>> &perms ){
        if( offset == num.size() ){
            perms.push_back( num );
            return ;
        }
        
        for( int i=offset; i < num.size(); i++){
            Swap( num, offset, i );
            PermuteHelp( num, offset+1, perms);
            Swap( num, offset, i );
        }
    }
    
    void Swap( vector<int> &num, int &a, int &b){
        if( a == b )
            return ;
            
        int temp=num[b];
        num[b] = num[a];
        num[a] = temp;
    }
};

File: SQL-AI/ProgrammingLanguage/Leetcode-Implementation/Python/Permutations.py
class Solution:
    # @param num, a list of integer
    # @return a list of lists of integers
    def permute(self, num):
        res = [] 
        self.permuteHelp(num, 0, res)
        return res
         
    def permuteHelp(self, num, start, res):
        if start == len(num) - 1:
            new_num = list(num)     # create a new list
            res.append( new_num )
            return
        
        self.permuteHelp( num, start + 1, res)  
        for i in range(start + 1, len(num)):
            num[i], num[start] = num[start], num[i]                 
            self.permuteHelp( num, start + 1, res)
            num[i], num[start] = num[start], num[i]                 







##################################
#
#   This class won't change num

#class Solution:
#    # @param num, a list of integer
#    # @return a list of lists of integers
#    def permute(self, num):
#        res = [] 
#        self.permuteHelp(num, 0, res)
#        print res
#        return res
#         
#    def permuteHelp(self, num, start, res):
#        if start == len(num) - 1:
#            #print num
#            res.append( num )
#            print res
#            return
#        
#        self.permuteHelp( num, start + 1, res)  
#        for i in range(start + 1, len(num)):
#            num[i], num[start] = num[start], num[i]                 
#            #print num
#            self.permuteHelp( num, start + 1, res)
#            num[i], num[start] = num[start], num[i]                 
                           
                           
if __name__ == "__main__":
    num = [1, 2, 3]
    inst = Solution()
    res = inst.permute( num )
    print res                            
